Use iterator of DispatchKeySet.

ailzhang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #44933 Align casing in test_dispatch with dispatch keys.
* #44556 Support Math keyword in native_functions.yaml.
* #44354 Add alias dispatch key Math.
* **#44682 Use iterator of DispatchKeySet.**
fixes #44731 
Differential Revision: [D23698387](https://our.internmc.facebook.com/intern/diff/D23698387)

Commit:
Use iterator of DispatchKeySet.

[ghstack-poisoned]

ailzhang commented on aten/src/ATen/core/dispatch/OperatorEntry.cpp:
```diff
@@ -220,7 +220,15 @@ void OperatorEntry::updateDispatchTableEntry_(const c10::Dispatcher& dispatcher,
-  for (auto k : c10::getRuntimeDispatchKeys(dispatch_key)) {
+  // Undefined isn't a runtime key but we have an entry for it in dispatchTable_ (which is not necessary).
+  // Update it separately if required.
+  // XXX: in fact I think we should just start from iter=1 in updateDispatchTableFull_.
```
cc: @ezyang Here's the special handling for `Undefined` I mentioned offline. ;)

ezyang replied:
OK.  Long term, I actually think having the entry for Undefined in the dispatch table is useful, as (1) it gives people place to specify functionality that should run when there are no dispatch keys, e.g., an empty TensorList argument and (2) it would let us remove the explicit error checking code in the dispatch hotpath, and so when no dispatch keys are available we just slide into the undefined handler which would then raise the error message.

But indeed, this is awkward if you want to represent RuntimeDispatchKeySet as a DispatchKeySet, as is done here. I think what we should probably do is just say that alias keys cannot validly include undefined tensor. (Except maybe the math key? Hmm...)

Commit:
Update on "Use iterator of DispatchKeySet."




[ghstack-poisoned]

ezyang reviewed (approved):
Thanks, much nicer.

ezyang commented on aten/src/ATen/core/dispatch/OperatorEntry.cpp:
```diff
@@ -220,7 +220,15 @@ void OperatorEntry::updateDispatchTableEntry_(const c10::Dispatcher& dispatcher,
+    updateDispatchTableEntry_(dispatcher, dispatch_key);
+    return;
+  }
+  for (auto k : c10::getRuntimeDispatchKeySet(dispatch_key)) {
```
Is getRuntimeDispatchKeySet a hot function? If it isn't we should make it raise an exception if it's passed Undefined.

ailzhang replied:
It's currently in hot path (that every dispatcher call go through it). We can remove it from hot path and only keeps it in precompute if we could move the autograd keyset constexpr to header file. But I'll need to run the ios binary size test on that so I'll do it in a followup PR. :D

ezyang replied:
Wait, what? That shouldn't be true.

```
(/home/ezyang/local/pytorch-tmp-env) [ezyang@devvm066.ash0 ~/local/pytorch-tmp] git grep getRuntimeDispatchKeySet
aten/src/ATen/core/LegacyTypeDispatch.h:    autograd_guard_(getRuntimeDispatchKeySet(DispatchKey::Autograd)) {
aten/src/ATen/templates/TensorBody.h:  return c10::impl::tls_local_dispatch_key_set().excluded_.isSupersetOf(c10::getRuntimeDispatchKeySet(DispatchKey::Autograd));
c10/core/DispatchKeySet.cpp:DispatchKeySet getRuntimeDispatchKeySet(DispatchKey t) {
c10/core/DispatchKeySet.cpp:  return k != DispatchKey::Undefined && getRuntimeDispatchKeySet(alias).has(k);
c10/core/DispatchKeySet.h:C10_API DispatchKeySet getRuntimeDispatchKeySet(DispatchKey t);
c10/core/DispatchKeySet.h://      we can remove this API and use c10::getRuntimeDispatchKeySet instead.
c10/core/DispatchKeySet.h:// getRuntimeDispatchKeySet(alias).has(DispatchKey::Undefind)
c10/core/DispatchKeySet.h:  return (s - getRuntimeDispatchKeySet(DispatchKey::Autograd)).highestPriorityTypeId();
c10/core/TensorImpl.h:    key_set_ = key_set_ - getRuntimeDispatchKeySet(DispatchKey::Autograd);
c10/core/impl/LocalDispatchKeySet.cpp:      raw_local_dispatch_key_set.excluded() | getRuntimeDispatchKeySet(DispatchKey::Autograd));
```

I don't think any of these are on dispatcher calls? (It looks like it does show up in autoNonVariableTypeMode). The autograd set should probably get constexpr'ed

ailzhang replied:
Yup `AutoNonVariableType` is the one I had in mind. Btw I updated the PR to separate autograd set to be a separate function returning a cheap copy of our constexpr (still keep the constexpr variable in cpp file so that we don't increase ios binary size). Please let me know if there's a better way to do it :D  Thanks!

ezyang replied:
I'm still confused. DispatchKeySet is one word. If it's all constexpr there shouldn't be any binary size problem putting it in a header. Am I missing something?

ezyang replied:
cc @bhosmer

Commit:
Update on "Use iterator of DispatchKeySet."


Differential Revision: [D23698387](https://our.internmc.facebook.com/intern/diff/D23698387)

[ghstack-poisoned]

Commit:
Update on "Use iterator of DispatchKeySet."


Differential Revision: [D23698387](https://our.internmc.facebook.com/intern/diff/D23698387)

[ghstack-poisoned]

ezyang commented on aten/src/ATen/core/LegacyTypeDispatch.h:
```diff
@@ -47,7 +47,7 @@ struct CAFFE2_API AutoNonVariableTypeMode {
   // TODO: Eliminate this parameter entirely
   AutoNonVariableTypeMode(bool enabled = true) :
-    autograd_guard_(getRuntimeDispatchKeySet(DispatchKey::Autograd)) {
+    autograd_guard_(getRuntimeAutogradKeySet()) {
```
Expecting an integer constant to get inlined here!

ezyang commented on c10/core/DispatchKeySet.cpp:
```diff
@@ -13,6 +13,7 @@ constexpr DispatchKeySet autograd_dispatch_keyset = DispatchKeySet({
+  TORCH_INTERNAL_ASSERT(t != DispatchKey::Undefined);
   switch (t) {
     case DispatchKey::Autograd:
       return autograd_dispatch_keyset;
```
Delete the Undefined case below now

ezyang commented:
This still LGTM but probably more things should be getting inlined than they are right now.

Commit:
Update on "Use iterator of DispatchKeySet."


Differential Revision: [D23698387](https://our.internmc.facebook.com/intern/diff/D23698387)

[ghstack-poisoned]
